xgcm/padding.py

```diff
@@ -0,0 +1,295 @@
+        ax_padding = padding[ax]
+        # translate padding and kwargs to xarray.pad syntax
+        ax_padding = _XGCM_BOUNDARY_KWARG_TO_XARRAY_PAD_KWARG[ax_padding]
+        if ax_padding == "constant":
+            kwargs = dict(constant_values=fill_value[ax])
+        else:
+            kwargs = dict()
+        da_padded = da_padded.pad({dim: widths}, ax_padding, **kwargs)
+    return da_padded
+
+
+def pad(
+    da: Union[xr.DataArray, Dict[str, xr.DataArray]],
+    grid: Grid,
+    boundary_width: Optional[Dict[str, Tuple[int, int]]] = None,
+    boundary: Optional[str] = None,
+    # fill_value: Optional[Union[float, Dict[str, float]]] = None,
+    **kwargs
+):
+    """
+    Pads the boundary of given arrays along given Axes, according to information in Axes.boundary.
+
+    Parameters
+    ----------
+    arrays :
+        Arrays to pad according to boundary and boundary_width.
+    boundary_width :
+        The widths of the boundaries at the edge of each array.
+        Supplied in a mapping of the form {axis_name: (lower_width, upper_width)}.
+    boundary : {None, 'fill', 'extend', 'extrapolate', dict}, optional
+        A flag indicating how to handle boundaries:
+        * None: Defaults to `periodic`
+        * 'periodic' : Wrap array along the specified axes
+        * 'fill':  Set values outside the array boundary to fill_value
+          (i.e. a Dirichlet boundary condition.)
+        * 'extend': Set values outside the array to the nearest array
+          value. (i.e. a limited form of Neumann boundary condition.)
+        * 'extrapolate': Set values by extrapolating linearly from the two
+          points nearest to the edge
+        Optionally a dict mapping axis name to separate values for each axis
+        can be passed.
+
+    fill_value :
+        The value to use in boundary conditions with `boundary='fill'`.
+        Optionally a dict mapping axis name to separate values for each axis
+        can be passed. Default is 0.
+    """
+    # TODO accept a general padding function like numpy.pad does as an argument to boundary
+
+    # TODO: boundary width should not really be optional here I think?
+    if not boundary_width:
+        raise ValueError("Must provide the widths of the boundaries")
+
+    # TODO rename this globally
+    padding = boundary
+    padding_width = boundary_width
+
+    padding_args = (da, grid, padding_width)
+    # Brute force approach for now. If any axis has connections use
+    # face connection method, this dispatches the simple pad stuff internally if needed.
+    if any([any(grid.axes[ax]._connections.keys()) for ax in grid.axes]):
+        da_padded = _pad_face_connections(*padding_args, padding=padding, **kwargs)
+    else:
+        # TODO: we need to have a better detection and checking here. For now just pipe everything else into the
+        # legacy cases
+
+        # Legacy cases
+        da_padded = _pad_basic(*padding_args, padding=padding, **kwargs)
```

Comment:
Sorry what exactly are the legacy cases that your code doesn't handle?

Reply:
So the legacy cases are with no face connections, handled in the way you and I did it using `xarray.pad`?

Reply:
That is maybe a bad naming. These are all the 'basic' cases we implemented before (e.g. everything xarray can handle). Does that make sense? The comments here were more for myself, thinking forward when we implement other options.